rust/datafusion/src/execution/datasource.rs

```diff
@@ -56,19 +84,40 @@ impl DataSource for CsvDataSource {
     }
 }
 
-#[derive(Serialize, Deserialize, Clone)]
-pub enum DataSourceMeta {
-    /// Represents a CSV file with a provided schema
-    CsvFile {
-        filename: String,
-        schema: Rc<Schema>,
-        has_header: bool,
-        projection: Option<Vec<usize>>,
-    },
-    /// Represents a Parquet file that contains schema information
-    ParquetFile {
-        filename: String,
-        schema: Rc<Schema>,
-        projection: Option<Vec<usize>>,
-    },
+pub trait DataSourceProvider {
+    fn schema(&self) -> &Arc<Schema>;
+    fn scan(&self, projection: &Option<Vec<usize>>) -> Rc<RefCell<DataSource>>;
+}
+
+/// Represents a CSV file with a provided schema
+pub struct CsvProvider {
+    filename: String,
+    schema: Arc<Schema>,
+    has_header: bool,
+}
+
+impl CsvProvider {
+    pub fn new(filename: &str, schema: &Schema, has_header: bool) -> Self {
+        Self {
+            filename: String::from(filename),
+            schema: Arc::new(schema.clone()),
+            has_header,
+        }
+    }
+}
+
+impl DataSourceProvider for CsvProvider {
+    fn schema(&self) -> &Arc<Schema> {
+        &self.schema
+    }
+
+    fn scan(&self, projection: &Option<Vec<usize>>) -> Rc<RefCell<DataSource>> {
+        Rc::new(RefCell::new(CsvDataSource::new(
+            &self.filename,
+            self.schema.clone(),
+            self.has_header,
+            projection,
+            1024 * 1024,
```

Comment:
nit: can this be a constant?

Reply:
I removed this hard-coded batch size and made it a parameter, so the tests specify the batch size now.